Ner

200kgsquat opened the pull request.

200kgsquat committed:
add train ner not clean code

200kgsquat committed:
add ner bert

ernestknurov commented on docs/ner_predictions.md:
```diff
@@ -0,0 +1,72 @@
+# BERT NER Predictions — Analysis
+
+This document summarizes predictions made by the BERT NER model (`bert-base-cased`) on the `jjzha/skillspan` dataset.
```
I would like to know training process details e.g. on how many epochs model were trained, how much data was used. Also I would like to see what model was used for baseline and what metrics were achieved. It's good for comparison.

ernestknurov commented on docs/ner_predictions.md:
```diff
@@ -0,0 +1,72 @@
+> Model correctly predicted all tokens as `O`.
+
+**Example 2:**  
+Sentence: "4 ) Work on a mix of front-end back-end and cloud technologies ."  
```
I don't think the described sentence has skills. Did you take this example from labeled dataset?

200kgsquat replied:
yes, it was random sample from my data

ernestknurov commented on src/app/core/ner/bert_ner.py:
```diff
@@ -0,0 +1,250 @@
+                pt = torch.exp(-ce_loss)
+                focal_loss = self.focal_alpha * (1 - pt) ** self.focal_gamma * ce_loss
+                mask = labels_flat != -100
+                loss = focal_loss[mask].mean() if mask.sum() > 0 else torch.tensor(0.0, device=logits.device, requires_grad=True)
```
focal loss - good idea 👍

ernestknurov commented on src/app/core/ner/bert_ner.py:
```diff
@@ -0,0 +1,250 @@
+                focal_loss = self.focal_alpha * (1 - pt) ** self.focal_gamma * ce_loss
+                mask = labels_flat != -100
+                loss = focal_loss[mask].mean() if mask.sum() > 0 else torch.tensor(0.0, device=logits.device, requires_grad=True)
+                return (loss, outputs) if return_outputs else loss
```
But why didn't you use the CRF with classification head? Without CRF you can get illegal or unlikely sequence of tags. Did you measure how often your model produce illegal sequence of tags?

200kgsquat replied:
If I get mostly illegal sequences, it will affect my metrics, for instance F1 span. But it doesn’t, so I can say BERT takes over CRF’s work.

ernestknurov reviewed (approved):
Please provide the PR description next time, e.g. what's changed.
Overall is good, but lacks the documentation details and CRF (or explanation that you don't need it)

200kgsquat committed:
add BERT NER performance tables

- Added BERT-NER classification report
- Added LR+CRF comparison
- Reformatted training section for readability

ernestknurov reviewed (approved):
👍
